File: js/ctci/3-stacks-and-queues/3.2-stack-min.js
class Stack {
  constructor() {
    this.data = [];
    this.size = 0;
  }

  push(value) {
    this.data.push(value);

    this.size++;
  }

  print() {
    console.log(this.data);
  }

  pop() {
    this.size--;
    return this.data.pop();
  }

  length() {
    return this.size;
  }

  peek() {
    return this.data[this.size - 1];
  }

  isEmpty() {
    return this.size === 0;
  }
}

class StackWithMin {
  constructor() {
    this.data = [];
    this.size = 0;

    this.minStack = new Stack();
  }

  push(value) {
    this.data.push(value);
    this.size++;
    if (this.minStack.isEmpty()) {
      this.minStack.push(value);
    }

    if (value < this.minStack.peek()) {
      this.minStack.push(value);
    }
  }

  print() {
    console.log(this.data);
  }

  min() {
    return this.minStack.peek();
  }

  pop() {
    this.size--;
    const pop = this.data.pop();
    if (pop === this.minStack.peek()) {
      this.minStack.pop();
    }
  }

  length() {
    return this.size;
  }

  peek() {
    return this.data[this.size - 1];
  }

  isEmpty() {
    return this.size === 0;
  }
}

const stack = new StackWithMin();
stack.push(3);
stack.push(2);
stack.push(4);
stack.push(5);
stack.push(1);
stack.push(7);

stack.pop();
console.log(stack.min()); // 1

stack.pop();
console.log(stack.min()); // 2

stack.pop();
stack.pop();
stack.pop();
console.log(stack.min()); // 3

File: python/ctci/3-stacks-and-queues/3.2.-stack-min.py


class Stack:
    def __init__(self):
        self.items = []

    def push(self, value):
        self.items.append(value)

    def pop(self):
        return self.items.pop()

    def peek(self):
        if self.items:
            return self.items[-1]
        return None

    def is_empty(self):
        return len(self.items) == 0


class MinStack(Stack):
    def __init__(self):
        super().__init__()
        self.minvals = Stack()

    def push(self, value):
        super().push(value)
        if self.minvals.is_empty():
            self.minvals.push(value)
        elif value <= self.minimum():
            self.minvals.push(value)

    def pop(self):
        value = super().pop()
        if value == self.minimum():
            self.minvals.pop()

        return value

    def minimum(self):
        return self.minvals.peek()


if __name__ == "__main__":
    min_stack = MinStack()
    min_stack.push(5)
    min_stack.push(4)
    min_stack.push(3)
    min_stack.pop()
    print(min_stack.minimum())

    # print(s.min())
